pkg/model/time_series_builder.go

```diff
@@ -25,59 +26,126 @@ func NewTimeSeriesBuilder(by ...string) *TimeSeriesBuilder {
+	s.labelBuf = lbs.BytesWithLabels(s.labelBuf, s.by...)
+	seriesKey := string(s.labelBuf)
+
 	pAnnotations := make([]*typesv1.ProfileAnnotation, 0, len(annotations.Keys))
 	for i := range len(annotations.Keys) {
 		pAnnotations = append(pAnnotations, &typesv1.ProfileAnnotation{
 			Key:   annotations.Keys[i],
 			Value: annotations.Values[i],
 		})
 	}
-	if !ok {
-		s.labelBuf = lbs.BytesWithLabels(s.labelBuf, s.by...)
-		labelsByString = string(s.labelBuf)
-		s.labelsByFingerprint[fp] = labelsByString
-		if _, ok := s.series[labelsByString]; !ok {
-			s.series[labelsByString] = &typesv1.Series{
-				Labels: lbs.WithLabels(s.by...),
-				Points: []*typesv1.Point{
-					{
-						Timestamp:   ts,
-						Value:       value,
-						Annotations: pAnnotations,
-					},
+
+	series, exists := s.series[seriesKey]
+	if !exists {
+		s.series[seriesKey] = &typesv1.Series{
+			Labels: lbs.WithLabels(s.by...),
+			Points: []*typesv1.Point{
+				{
+					Timestamp:   ts,
+					Value:       value,
+					Annotations: pAnnotations,
 				},
-			}
-			return
+			},
 		}
+	} else {
+		series.Points = append(series.Points, &typesv1.Point{
+			Timestamp:   ts,
+			Value:       value,
+			Annotations: pAnnotations,
+		})
+	}
+
+	if profileID != "" {
+		if s.exemplarBuilders[seriesKey] == nil {
+			s.exemplarBuilders[seriesKey] = NewExemplarBuilder()
+		}
+		s.exemplarBuilders[seriesKey].Add(fp, lbs, ts, profileID, uint64(value))
 	}
-	series := s.series[labelsByString]
-	series.Points = append(series.Points, &typesv1.Point{
-		Timestamp:   ts,
-		Value:       value,
-		Annotations: pAnnotations,
-	})
 }
 
+// Build returns the time series without exemplars.
 func (s *TimeSeriesBuilder) Build() []*typesv1.Series {
 	return s.series.normalize()
 }
 
+// BuildWithExemplars returns the time series with exemplars attached.
+func (s *TimeSeriesBuilder) BuildWithExemplars() []*typesv1.Series {
+	series := s.series.normalize()
+	s.attachExemplars(series)
+	return series
+}
+
+// attachExemplars attaches exemplars from ExemplarBuilders to the corresponding points.
+func (s *TimeSeriesBuilder) attachExemplars(series []*typesv1.Series) {
+	// Create a map from seriesKey to series for fast lookup
+	seriesMap := make(map[string]*typesv1.Series)
+	for _, ser := range series {
+		seriesKey := string(Labels(ser.Labels).BytesWithLabels(nil, s.by...))
+		seriesMap[seriesKey] = ser
+	}
+
+	for seriesKey, exemplarBuilder := range s.exemplarBuilders {
+		ser, found := seriesMap[seriesKey]
+		if !found {
+			continue
+		}
+
+		exemplars := exemplarBuilder.Build()
+		if len(exemplars) == 0 {
+			continue
+		}
+
+		exemplarsByTimestamp := make(map[int64][]*typesv1.Exemplar)
+		for _, ex := range exemplars {
+			exemplarsByTimestamp[ex.Timestamp] = append(exemplarsByTimestamp[ex.Timestamp], ex)
+		}
+
+		for _, point := range ser.Points {
+			if exs, found := exemplarsByTimestamp[point.Timestamp]; found {
+				point.Exemplars = exs
+			}
+		}
+	}
```

Comment:
As exemplars are already time ordered, I think we could get rid of that extra map.